[WIP] Add examples of usage of vector integration (#20021)
sympy/sympy: doc/src/modules/vector/vector_integration.rst

>>> from sympy import sin, cos, exp, pi
>>> from sympy.vector import CoordSys3D, ParametricRegion, vector_integrate
>>> from sympy.abc import r, x, y, z, theta, phi
>>> C = CoordSys3D('C')
    
Calculation of Perimeter, Surface Area and Volume
=================================================

To calculate the perimeter of circle, we need to define a circle. Let's define it using its parametric equation.

>>> circle = ParametricRegion((4*cos(theta), 4*sin(theta)), (theta, 0, 2*pi))

To calculate the perimeter of the circle, we can integrate it over a scalar field of unit magnitude.

>>> vector_integrate(1, circle)
8*pi

Suppose a user wants to calculate the perimeter of triangle. Determining the parametric representation of triangle can be difficult. Instead, the user can use an object of :class:`~sympy.geometry.polygon.Polygon` class in the geometry module.

>>> from sympy.geometry import Point, Polygon 
>>> triangle = Polygon(Point(1, 2), (3, 5), (1,6))
>>> vector_integrate(1, triangle)
sqrt(5) + sqrt(13) + 4

To define a solid sphere, we need to use three parameters (r, theta and phi).

>>> sphere = ParametricRegion((r*sin(phi)*cos(theta),r*sin(phi)*sin(theta), r*cos(phi)),

> ... (theta, 0, pi), (phi, 0, pi), (r, 0, 3))

Reviewer: maybe `theta` should range from 0 to `2*pi`?

Author: Yes, will fix it.

>>> vector_integrate(1, sphere)

> -18*pi

Reviewer: why is this result negative? I would indeed expect `18*pi`, the volume of the half-sphere, as you have defined `theta` to reach `pi` (instead of `2*pi`).

Author: In ParametricIntegral class

```python
variables = cls._bounds_case(parametricregion.limits)
coeff = Matrix(parametricregion.definition).jacobian(variables).det()
integrand = simplify(parametricfield*coeff)
l = [(var, parametricregion.limits[var][0], parametricregion.limits[var][1]) for var in variables]
result = integrate(integrand, *l)
```

The volume element requires calculating the jacobian. So the sign depends on the order of variables used in the calculation of jacobian. The order is determined by the bounds_case function which is necessary for interdependent limits. So the sign cannot be changed without breaking test cases.

How about using `abs` function?
```
>>> abs(vector_integrate(1, sphere))
18*pi
```

Author: For example, 
```python
sphere1 = ParametricRegion((r*sin(phi)*cos(theta),r*sin(phi)*sin(theta), r*cos(phi)),\
                                                (theta, 0, pi), (phi, 0, pi), (r, 0, 3))
sphere2 = ParametricRegion((r*sin(phi)*cos(theta),r*sin(phi)*sin(theta), r*cos(phi)),\
                                                  (r, 0, 3), (theta, 0, pi), (phi, 0, pi))
```
In the present implementation, vector_integrate(1, sphere1) and vector_integrate(1, sphere2) returns result with same sign. 
Is this desired or should the order matter?

=================================================

To calculate the perimeter of circle, we need to define a circle. Let's define it using its parametric equation.

>>> param_circle = ParametricRegion((4*cos(theta), 4*sin(theta)), (theta, 0, 2*pi))

We can also define circle using its implicit equation.

>>> implicit_circle = ImplicitRegion((x, y), x**2 + y**2 - 4)

Perimeter of a figure is equal to the absolute value of its integral over a unit scalar field. 

>>> vector_integrate(1, param_circle)
8*pi
>>> vector_integrate(1, implicit_circle)
8*pi

Suppose a user wants to calculate the perimeter of triangle. Determining the parametric representation of triangle can be difficult. Instead, the user can use an object of :class:`~sympy.geometry.polygon.Polygon` class in the geometry module.

>>> from sympy.geometry import Point, Polygon 
>>> triangle = Polygon(Point(1, 2), (3, 5), (1,6))
>>> vector_integrate(1, triangle)
sqrt(5) + sqrt(13) + 4

To define a solid sphere, we need to use three parameters (r, theta and phi). 

>>> solidsphere = ParametricRegion((r*sin(phi)*cos(theta),r*sin(phi)*sin(theta), r*cos(phi)),
... (phi, 0, pi), (theta, 0, 2*pi), (r, 0, 3))

> >>> abs(vector_integrate(1, solidsphere))

Reviewer: is it a problem in the algorithm that you get negative results? Why is `vector_integrate` returning a negative result?

Author: This is due to the order of variables used in the calculation of Jacobian. The order is decided by function `bounds_case`. `bounds_case` is required for dealing with situations where bounds of variables are interdependent.

Please see my comment above on this.

Author: I have made a PR https://github.com/sympy/sympy/pull/20044 for this.

Reviewer: OK, it's merged. Can you remove `abs( )` now?

Author: Yes.

36*pi

Calculation of mass of a body
=============================

Consider a triangular lamina 𝑅  with vertices (0,0), (0, 5), (5,0) and with density :math:`\rho(x, y) = xy\:kg/m^2`. Find the total mass.

> >>> triangle = ParametricRegion((x, y), (x, 0, 5), (y, 0, 5 - x))

Reviewer: does this work with the `polygon` object as well?

Author: No. `vector_integrate` will perform line integral over the sides of the polygon instead of calculating the area enclosed.

Reviewer: OK, maybe that's an issue for some future work.

Author: Yes. We can add new classes in geometry module for representing such regions.

Author: I will open an issue for this then.

Reviewer: It would require an algorithm to decompose the polygon into triangles (or pyramids in the 3D case). Well, maybe for GSoC 2021 :)

Author: Yes, for polygons it will be difficult. But I think adding a class to represent a triangular area will be easier.

================================
Applications of Vector Integrals
================================

To integrate a scalar or vector field over a region, we have to first define a region. SymPy provides three methods for defining a region. 

1. Using Parametric Equations.

2. Using Implicit Equation.

3. Using objects of geometry module.

> The vector_integrate method is used to integrate scalar or vector field over any type of region. It automatically determines the type of integration (line, surface or volume) depending on the nature of the object.

Reviewer 2: ```suggestion
The :func:`~sympy.vector.vector_integrate` function is used to integrate scalar or vector field over any type of region. It automatically determines the type of integration (line, surface or volume) depending on the nature of the object.
```

Author: Thanks for the suggestions.

Importing necessary classes and functions:

>>> from sympy.vector import pi
>>> from sympy.vector import ParametricRegion, vector_integrate
>>> from sympy.abc import r, x, y, z, theta, phi
>>> C = CoordSys3D('C')
    
Calculation of Perimeter, Surface Area and Volume
=================================================

> To calculate the perimeter of circle, we need to define a circle, Lets define it using its parametric equation.

Reviewer 2: ```suggestion
To calculate the perimeter of circle, we need to define a circle. Let's define it using its parametric equation.
```

>>> circle = ParametricRegion((4*cos(theta), 4*sin(theta)), (theta, 0, 2*pi))

To calculate the perimeter of the circle, we can integrate it over a scalar field of unit magnitude.

>>> vector_integrate(1, circle)
8*pi

> Suppose a user wants to calculate the perimeter of triangle. Determining the parametric representation of triangle can be difficult. Instead, the user can use an object of Polygon class in the geometry module.

Reviewer 2: ```suggestion
Suppose a user wants to calculate the perimeter of triangle. Determining the parametric representation of triangle can be difficult. Instead, the user can use an object of :class:`~sympy.geometry.Polygon` class in the geometry module.
```
